src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -0,0 +1,1521 @@
+     let gslot ← LLVM.getOrAddGlobal (← getLLVMModule) fcname retty
+     LLVM.buildLoad2 builder retty gslot
+  else
+    let argtys ← decl.params.mapM (fun p => do toLLVMType p.ty)
+    let fnty ← LLVM.functionType retty argtys
+    LLVM.getOrAddFunction (← getLLVMModule) fcname fnty
+
+def emitPartialApp (builder : LLVM.Builder llvmctx) (z : VarId) (f : FunId) (ys : Array Arg) : M llvmctx Unit := do
+  let decl ← getDecl f
+  let fv ← getOrAddFunIdValue builder f
+  let arity := decl.params.size;
+  let (_zty, zslot) ← emitLhsSlot_ z
+  let fv_voidptr ← LLVM.buildPointerCast builder fv (← LLVM.voidPtrType llvmctx)
+  let zval ← callLeanAllocClosureFn builder fv_voidptr
+                                    (← constIntUnsigned arity)
+                                    (← constIntUnsigned ys.size)
+  LLVM.buildStore builder zval zslot
+  ys.size.forM fun i => do
+    let (yty, yslot) ← emitArgSlot_ builder ys[i]!
+    let yval ← LLVM.buildLoad2 builder yty yslot
+    callLeanClosureSetFn builder zval (← constIntUnsigned i) yval
+
+def emitApp (builder : LLVM.Builder llvmctx) (z : VarId) (f : VarId) (ys : Array Arg) : M llvmctx Unit := do
+  if ys.size > closureMaxArgs then do
+    let aargs ← LLVM.buildAlloca builder (← LLVM.arrayType (← LLVM.voidPtrType llvmctx) (UInt64.ofNat ys.size)) "aargs"
+    for i in List.range ys.size do
+      let (yty, yv) ← emitArgVal builder ys[i]!
+      let aslot ← LLVM.buildInBoundsGEP2 builder yty aargs #[← constIntUnsigned 0, ← constIntUnsigned i] s!"param_{i}_slot"
+      LLVM.buildStore builder yv aslot
+    let fnName :=  s!"lean_apply_m"
+    let retty ← LLVM.voidPtrType llvmctx
+    let args := #[← emitLhsVal builder f, ← constIntUnsigned ys.size, aargs]
+    -- '1 + ...'; '1' for the fn and 'args' for the arguments
+    let argtys := #[← LLVM.voidPtrType llvmctx]
+    let fn ← getOrCreateFunctionPrototype (← getLLVMModule) retty fnName argtys
+    let fnty ← LLVM.functionType retty argtys
+    let zv ← LLVM.buildCall2 builder fnty fn args
+    emitLhsSlotStore builder z zv
+  else do
+
+    let fnName :=  s!"lean_apply_{ys.size}"
+    let retty ← LLVM.voidPtrType llvmctx
+    let args : Array (LLVM.Value llvmctx) := #[← emitLhsVal builder f] ++ (← ys.mapM (fun y => Prod.snd <$> (emitArgVal builder y)))
+    -- '1 + ...'; '1' for the fn and 'args' for the arguments
+    let argtys := (List.replicate (1 + ys.size) (← LLVM.voidPtrType llvmctx)).toArray
+    let fn ← getOrCreateFunctionPrototype (← getLLVMModule) retty fnName argtys
+    let fnty ← LLVM.functionType retty argtys
+    let zv ← LLVM.buildCall2 builder fnty fn args
+    emitLhsSlotStore builder z zv
+
+def emitFullApp (builder : LLVM.Builder llvmctx)
+  (z : VarId) (f : FunId) (ys : Array Arg) : M llvmctx Unit := do
+  let (__zty, zslot) ← emitLhsSlot_ z
+  let decl ← getDecl f
+  match decl with
+  | Decl.extern _ ps retty extData =>
+     let zv ← emitExternCall builder f ps extData ys retty
+     LLVM.buildStore builder zv zslot
+  | _ =>
```

Comment:
Is there a particular reason for the `_` here? In general I feel like adding wildcards when you don't need them can be dangerous, if we end up having a new `Decl` variant in the future for some reason we might overlook something here because the compiler did not tell us.

Reply:
Fixed here. In some other places, I use `_` where there are quite a few variants to be handled (eg. `IR.isIntegerType`). Lean does not have [or-patterns](https://v2.ocaml.org/manual/patterns.html#sss:pat-or), does it? If it does, then I could write the code as:

```lean
match scrutinee with
| useful => ...
| useless1 | useless2 | ... | uselessN => throw ... 
```

Without this construct, it would lead to a bunch of duplication of the RHS:

```lean
match scrutinee with
| useful => ...
| useless1 => throw ... 
| useless2 => throw ...  -- duplicated
| ...  -- duplicated
| uselessN => throw ... -- duplicated 
```

Reply:
> Lean does not have [or-patterns](https://v2.ocaml.org/manual/patterns.html#sss:pat-or), does it?

It does, using that exact syntax